### DISCUSSION TITLE
LVGL 9.1??

### DISCUSSION THREAD

**Discussion by GeorgeIoak** [Westcott1/Waveshare-ESP32-S3-Touch-LCD-4.3-and-Arduino#5] [General]
I came across this repo while struggling with my ESP32-S3 board and trying to get my ILI9488 SPI working.

It seems others have similar issues with TFT_eSPI and I too had to use LovyanGFX in order to get anything to work. But for some reason I get nothing but compile errors despite all my efforts to use LVGL 9.1.

Perhaps we could share efforts in trying to get LVGL to work?

**Comment by Westcott1** [comment]
Hi George,
Good to hear from you!
Yes, by all means let's share.
Can you share some of your errors with me?
I'll post an LVGL 9 example shortly.

**Reply by GeorgeIoak** [reply]
- I haven't used LVGL before, only TFT_eSPI.
- My older projects are based on the original ESP32 but recent projects I have switched to using ESP32-S3.
- I typically use PlatformIO but sometimes experiment/use Arduino 2.x IDE
- I haven't done a ESP-IDF project but have it installed and have experimented
- I have some development boards but usually design and build my own hardware

OK, I thought I'd get that out first. I know there's some big differences between LVGL 8.3 and 9.1 but since I hadn't used LVGL before I thought I would start with 9.1 so anything I learn will be useful in a few years.

I kept getting device panics when trying to use TFT_eSPI and my ILI9488 SPI 3.5" display. My board with ESP32 works perfectly but both PlatformIO and Arduino IDE failed. On a whim I tried LovyanGFX in Arduino and that worked, even with the FT6236 based capacitive touch screen.

I tried using ESP-IDF but even that was giving me compile errors with a blank project but TBH there seems to be so many options for LCD drivers I'm at a loss as to which ones should be used.

Since I'm most comfortable using PlatformIO with Arduino Framework that's my goal although there appears to be some issues with the PIO/Espressif relationship so this might be a good time to try and make the switch to ESP-IDF based projects.

**Comment by GeorgeIoak** [comment]
I tried downgrading my Arduino Core to 2.0.17 but I still had problems with panics. After some digging I stumbled across adding `#define USE_HSPI_PORT` to my user setup. I now have a working LVGL 9.1 example running and after some testing I plan on switching back to Arduino Core 3.

I also have this example working in PlatformIO but PIO is behind and still uses Arduino Core 2.0.14

I also have a discussion on TFT_eSPI at https://github.com/Bodmer/TFT_eSPI/discussions/3392#discussioncomment-9989889

**Comment by ulijoergens** [comment]
I managed to get my LVGL stuff working with some minor adjustments. The following example is based on your example code here, Lovyan GFX 1.1.12, LVGL 9.1.0 and esp32 2.0.17 (I had trouble with esp32 3.0 so I didn't upgrade to the version yet).
The order of the includes is important otherwise it won't compile.
The example displays the LGFX text (so you see at least something in case lvgl does fail for some reason).
It displays a label and a button. Clicking the button writes some text to the serial console.
Remeber to copy lvgl_config.h to the Arduine library folder (follow the LVGL documentation for details).

Copy the following into your sketch:

```
#include <Arduino.h>
#include <ESP_IOExpander_Library.h>
#define LGFX_USE_V1
#include <LovyanGFX.hpp>
#include "Pins.h"
#include "Wire.h"
#include "esp32-hal-psram.h"
#include <lvgl.h>

ESP_IOExpander *expander;
// Extender Pin define
#define TP_RST 1
#define LCD_BL 2
#define LCD_RST 3
#define SD_CS 4
#define USB_SEL 5

// I2C Pin define
#define I2C_MASTER_NUM 0
#define I2C_MASTER_SDA_IO 8
#define I2C_MASTER_SCL_IO 9

void ExpanderInit() {
  Serial.println("IO expander init...");
  Wire.begin(I2C_MASTER_SDA_IO, I2C_MASTER_SCL_IO);
  expander = new ESP_IOExpander_CH422G((i2c_port_t)I2C_MASTER_NUM, ESP_IO_EXPANDER_I2C_CH422G_ADDRESS_000);
  expander->init();
  expander->begin();
  expander->multiPinMode(TP_RST | LCD_BL | LCD_RST | SD_CS | USB_SEL, OUTPUT);
  BLblink();
//  expander->multiPinMode(LCD_BL | USB_SEL, OUTPUT);
//  expander->pinMode( LCD_BL, OUTPUT);
//  expander->digitalWrite(USB_SEL, LOW);
//  expander->multiDigitalWrite(TP_RST | LCD_BL | LCD_RST | SD_CS | USB_SEL, HIGH);
//  expander->multiDigitalWrite(TP_RST | LCD_BL | LCD_RST | SD_CS | USB_SEL, LOW);
}

void BLset(byte state){
    expander->digitalWrite(LCD_BL, state);
}

void BLblink(){
  //Turn off backlight
  BLset(LOW);
  delay(100);
  BLset(HIGH);
  delay(100);
}

LGFX gfx;

const char* title = "Hello World";

void LgfxInit() {
  Serial.println("Lgfx init...");

  gfx.init();

  BLset(HIGH);
  //  gfx.setBrightness(200);
  //  gfx.setRotation(2);

  gfx.fillScreen(0x003030);
  gfx.fillRect(0, 0, 800, 70, 0x000028);
  gfx.setTextColor(TFT_WHITE);
  gfx.setTextSize(4);
  gfx.setCursor(250, 20);
  gfx.println(title);

  gfx.setTextColor(TFT_GREEN);
  gfx.setTextSize(3);
}

const unsigned int screenWidth = TFT_HOR_RES;
const unsigned int screenHeight = TFT_VER_RES;
const unsigned int lvBufferSize = screenWidth * screenHeight / 10 * (LV_COLOR_DEPTH / 8);
uint8_t lvBuffer[lvBufferSize];

// Only if GFX does the touch handling
void LgfxDoTouch() {
  int32_t x, y;
  if (gfx.getTouch(&x, &y)) {
    Serial.printf("X:%d Y:%d\n", x, y);
    gfx.fillCircle(x, y, 15, TFT_WHITE);
  }
}

static void button_event_handler(lv_event_t * e) {
    lv_event_code_t code = lv_event_get_code(e);
//    lv_obj_t * target = (lv_obj_t*) lv_event_get_target(e);
    const char *text = (const char *) lv_event_get_user_data(e);

    if(code == LV_EVENT_CLICKED) {
      Serial.println(text);
    }
  };

void setup() {
  Serial.begin(115200);
  while (!Serial)
    ;

  Serial.println("Initialising...");

  //  TouchInit();
  delay(200);
  ExpanderInit();
  LgfxInit();
  Serial.println("Lgfx initialised");
  lv_init();
  static auto *lvDisplay = lv_display_create(TFT_HOR_RES,TFT_VER_RES); //screenWidth, screenHeight);
  lv_display_set_color_format(lvDisplay, LV_COLOR_FORMAT_RGB565);
  lv_display_set_flush_cb(lvDisplay, [](lv_display_t* display, const lv_area_t* area, unsigned char* data) {
          uint32_t w = lv_area_get_width(area);
          uint32_t h = lv_area_get_height(area);
          lv_draw_sw_rgb565_swap(data, w*h);
          gfx.pushImage(area->x1, area->y1, w, h, (uint16_t*)data);
          lv_display_flush_ready(display);
      });
  lv_display_set_buffers(lvDisplay, lvBuffer, nullptr, lvBufferSize, LV_DISPLAY_RENDER_MODE_PARTIAL);

  gfx.print("Input device...");
  static auto *lvInput = lv_indev_create();
  lv_indev_set_type(lvInput, LV_INDEV_TYPE_POINTER);
  lv_indev_set_read_cb(lvInput, [](lv_indev_t* dev, lv_indev_data_t* data) {
    uint16_t touchX, touchY;

    data->state = LV_INDEV_STATE_REL;

    if (gfx.getTouch(&touchX, &touchY))
    {
      data->state = LV_INDEV_STATE_PR;
      data->point.x = touchX;
      data->point.y = touchY;
    }
  });

  lv_obj_t *label = lv_label_create( lv_scr_act() );
  lv_label_set_text( label, "Hello World, here is LVGL!" );
  lv_obj_align( label, LV_ALIGN_CENTER, 0, 0 );
  lv_obj_t *button = lv_button_create( lv_scr_act() );
  lv_obj_t *buttonLabel = lv_label_create( button );
  lv_label_set_text( buttonLabel, "Click here..." );
  lv_obj_align_to( button,label, LV_ALIGN_OUT_BOTTOM_MID, 0, 0 );
  const char *event_text = "Click!";
  lv_obj_add_event_cb(button, button_event_handler, LV_EVENT_ALL, (void *) event_text);

}

void loop() {
    lv_task_handler(); 
    lv_tick_inc(5);
    delay(5);
}

```

Hope it helps!
